File: Algo/Algorithm/Questions/Arrays/11.longestPeak/main.js
// O(n) time | O(1) space - where n is the length of the array
function longestPeak(array) {
  let longestPeak = 0;
  let i = 1;

  while (i < array.length - 1) {
    const peak = array[i] > array[i - 1] && array[i] > array[i + 1];

    if (!peak) {
      i += 1;
      // eslint-disable-next-line no-continue
      continue;
    }

    let leftIndx = i - 2;
    while (leftIndx >= 0 && array[leftIndx] < array[leftIndx + 1]) {
      leftIndx -= 1;
    }

    let rightIndx = i + 2;
    while (rightIndx < array.length && array[rightIndx] < array[rightIndx - 1]) {
      rightIndx += 1;
    }

    const currentPeak = rightIndx - leftIndx - 1;
    longestPeak = Math.max(currentPeak, longestPeak);
    i = rightIndx;
  }

  return longestPeak;
}

File: Algo/Algorithm/Questions/Arrays/11.longestPeak/main.cpp
// O(n) time | O(1) space - where n is the length of the input array
#include <iostream>
#include <vector>

using namespace std;

int longestPeak(vector<int> array)
{
  int peak = 0;
  int i = 1;
  while (i < int(array.size() - 1))
  {
    bool isPeak = (array[i] > array[i - 1]) && (array[i] > array[i + 1]);
    if (!isPeak)
    {
      i++;
      continue;
    }

    int left = i - 2;
    while (left >= 0 && array[left] < array[left + 1])
    {
      left--;
    }
    int right = i + 2;
    while (right < array.size() && array[right] < array[right - 1])
    {
      right++;
    }

    int currentPeak = right - left - 1;
    peak = max(peak, currentPeak);
    i = right;
  }

  return peak;
}

int main()
{

  vector<int> array = {1, 2, 3, 3, 4, 0, 10, 6};
  cout << longestPeak(array) << endl;
  return 0;
}